modules/block_motd.cpp

```diff
@@ -44,21 +52,54 @@ class CBlockMotd : public CModule {
         const CString sCmd = sLine.Token(1);
 
         if ((sCmd == "375" /* begin of MOTD */ || sCmd == "372" /* MOTD */) &&
-            !m_bTemporaryAcceptMotd)
+            !ShouldTemporarilyAcceptMotd())
             return HALT;
 
         if (sCmd == "376" /* End of MOTD */) {
-            if (!m_bTemporaryAcceptMotd) {
+            if (!ShouldTemporarilyAcceptMotd()) {
                 sLine = sLine.Token(0) + " 422 " + sLine.Token(2) +
                         " :MOTD blocked by ZNC";
             }
-            m_bTemporaryAcceptMotd = false;
+            StopTemporarilyAcceptingMotd();
         }
         return CONTINUE;
     }
 
+    void OnIRCDisconnected() override {
+        StopTemporarilyAcceptingMotd();
+    }
+
+    bool ShouldTemporarilyAcceptMotd() const {
+        for (CIRCSock* pSock : m_vTemporaryAcceptedMotdSocks) {
+            if (pSock == GetNetwork()->GetIRCSock()) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    void TemporarilyAcceptMotd() {
+        if (ShouldTemporarilyAcceptMotd()) {
+            return;
+        }
+
+        m_vTemporaryAcceptedMotdSocks.push_back(GetNetwork()->GetIRCSock());
+    }
+
+    void StopTemporarilyAcceptingMotd() {
+        for (vector<CIRCSock*>::iterator it = m_vTemporaryAcceptedMotdSocks.begin();
+            it != m_vTemporaryAcceptedMotdSocks.end(); ++it)
+        {
+            if (*it == GetNetwork()->GetIRCSock()) {
+                m_vTemporaryAcceptedMotdSocks.erase(it);
+                return;
+            }
+        }
+    }
+
   private:
-    bool m_bTemporaryAcceptMotd = false;
+    vector<CIRCSock *> m_vTemporaryAcceptedMotdSocks;
```

Comment:
How about turning this into a `set` and using `count(GetNetwork()->GetIRCSock() > 0` and `erase(GetNetwork()->GetIRCSock()` instead of iterating through the vector?